File: apps/web/src/lib/realtime/protocol.test.ts
// eslint-disable-next-line import-x/no-unresolved -- "bun:test" is a virtual Bun built-in with no resolvable file path
import { describe, expect, test } from "bun:test";

import {
  eventEnvelopeSchema,
  parseIncomingMessage,
  parseRoomKeyParts,
  REAUTH_REQUIRED_CLOSE_CODE,
  roomKeySchema,
  systemMessageSchema,
} from "./protocol";

const ENVELOPE = {
  id: "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  type: "grades.published",
  room: "school:123:role:STUDENT",
  payload: { submissionId: "s-1", studentId: "u-1" },
  publishedAt: "2026-07-09T12:00:00.000Z",
};

describe("parseRoomKeyParts", () => {
  test("parses all three room kinds", () => {
    expect(parseRoomKeyParts("school:123")).toEqual({ kind: "school", schoolId: "123" });
    expect(parseRoomKeyParts("school:123:role:STUDENT")).toEqual({
      kind: "role",
      schoolId: "123",
      role: "STUDENT",
    });
    expect(parseRoomKeyParts("school:123:user:u-1")).toEqual({
      kind: "user",
      schoolId: "123",
      userId: "u-1",
    });
  });

  test("rejects malformed and cross-kind keys", () => {
    for (const value of [
      "",
      "school:",
      "tenant:123",
      "school:123:role:",
      "school:123:user:",
      "school:123:group:g-1",
      "school:123:role:STUDENT:extra",
    ]) {
      expect(parseRoomKeyParts(value)).toBeUndefined();
      expect(roomKeySchema.safeParse(value).success).toBe(false);
    }
  });
});

describe("wire schemas", () => {
  test("event envelope validates", () => {
    expect(eventEnvelopeSchema.safeParse(ENVELOPE).success).toBe(true);
  });

  test("system messages validate", () => {
    for (const message of [
      { type: "system.joined", room: "school:123" },
      { type: "system.left", room: "school:123" },
      { type: "system.error", message: "cannot join a room outside your own school" },
      { type: "system.reauth_required", reason: "token expired" },
    ]) {
      expect(systemMessageSchema.safeParse(message).success).toBe(true);
    }
  });

  test("reauth close code is 4401", () => {
    expect(REAUTH_REQUIRED_CLOSE_CODE).toBe(4401);
  });
});

describe("parseIncomingMessage", () => {
  test("parses an event envelope frame", () => {
    const message = parseIncomingMessage(JSON.stringify(ENVELOPE));
    expect(message).toEqual(ENVELOPE);
  });

  test("parses a system message frame", () => {
    const message = parseIncomingMessage(JSON.stringify({ type: "system.error", message: "nope" }));
    expect(message).toEqual({ type: "system.error", message: "nope" });
  });

  test("drops malformed JSON", () => {
    expect(parseIncomingMessage("not json")).toBeUndefined();
  });

  test("drops well-formed JSON that is not a known message", () => {
    expect(parseIncomingMessage(JSON.stringify({ type: "system.unknown" }))).toBeUndefined();
    expect(parseIncomingMessage(JSON.stringify({ ...ENVELOPE, id: "not-a-uuid" }))).toBeUndefined();
  });
});

File: apps/mobile/test/core/realtime/protocol_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:studafy_mobile/src/core/realtime/protocol.dart';

void main() {
  group('parseRoomKeyParts', () {
    test('parses a school room', () {
      final parsed = parseRoomKeyParts('school:123');
      expect(parsed, isA<ParsedRoomKey>());
      expect(parsed!.kind, RoomKind.school);
      expect(parsed.schoolId, '123');
    });

    test('parses a role room', () {
      final parsed = parseRoomKeyParts('school:123:role:STUDENT');
      expect(parsed!.kind, RoomKind.role);
      expect(parsed.schoolId, '123');
      expect(parsed.role, 'STUDENT');
    });

    test('parses a user room', () {
      final parsed = parseRoomKeyParts('school:123:user:user-1');
      expect(parsed!.kind, RoomKind.user);
      expect(parsed.schoolId, '123');
      expect(parsed.userId, 'user-1');
    });

    test('rejects an empty schoolId', () {
      expect(parseRoomKeyParts('school:'), isNull);
    });

    test('rejects a middle segment other than role/user', () {
      expect(parseRoomKeyParts('school:123:class:STUDENT'), isNull);
    });

    test('rejects a trailing empty segment', () {
      expect(parseRoomKeyParts('school:123:role:'), isNull);
    });

    test('rejects a foreign prefix', () {
      expect(parseRoomKeyParts('district:123'), isNull);
    });

    test('rejects the wrong number of segments', () {
      expect(parseRoomKeyParts('school:123:role'), isNull);
      expect(parseRoomKeyParts('school:123:role:STUDENT:extra'), isNull);
    });
  });

  group('EventEnvelope.tryParse', () {
    final validJson = <String, dynamic>{
      'id': '3fa85f64-5717-4562-b3fc-2c963f66afa6',
      'type': 'grades.published',
      'room': 'school:123:role:STUDENT',
      'payload': {'text': 'hello'},
      'publishedAt': '2026-07-09T12:00:00.000Z',
    };

    test('parses a valid envelope', () {
      final envelope = EventEnvelope.tryParse(validJson);
      expect(envelope, isNotNull);
      expect(envelope!.id, '3fa85f64-5717-4562-b3fc-2c963f66afa6');
      expect(envelope.type, 'grades.published');
      expect(envelope.room, 'school:123:role:STUDENT');
      expect(envelope.payload, {'text': 'hello'});
      expect(envelope.publishedAt, DateTime.parse('2026-07-09T12:00:00.000Z'));
    });

    test('rejects a non-UUID id', () {
      expect(
        EventEnvelope.tryParse({...validJson, 'id': 'not-a-uuid'}),
        isNull,
      );
    });

    test('rejects an empty type', () {
      expect(EventEnvelope.tryParse({...validJson, 'type': ''}), isNull);
    });

    test('rejects an invalid room key', () {
      expect(EventEnvelope.tryParse({...validJson, 'room': 'nope'}), isNull);
    });

    test('rejects an unparsable publishedAt', () {
      expect(
        EventEnvelope.tryParse({...validJson, 'publishedAt': 'not-a-date'}),
        isNull,
      );
    });

    test('rejects a missing field', () {
      final withoutId = Map<String, dynamic>.of(validJson)..remove('id');
      expect(EventEnvelope.tryParse(withoutId), isNull);
    });
  });

  group('parseIncomingMessage', () {
    test('parses system.joined', () {
      final message = parseIncomingMessage(
        '{"type":"system.joined","room":"school:123"}',
      );
      expect(message, isA<SystemJoined>());
      expect((message as SystemJoined).room, 'school:123');
    });

    test('parses system.left', () {
      final message = parseIncomingMessage(
        '{"type":"system.left","room":"school:123"}',
      );
      expect(message, isA<SystemLeft>());
    });

    test('parses system.error', () {
      final message = parseIncomingMessage(
        '{"type":"system.error","message":"nope"}',
      );
      expect(message, isA<SystemErrorMessage>());
      expect((message as SystemErrorMessage).message, 'nope');
    });

    test('parses system.reauth_required', () {
      final message = parseIncomingMessage(
        '{"type":"system.reauth_required","reason":"token expired"}',
      );
      expect(message, isA<SystemReauthRequired>());
      expect((message as SystemReauthRequired).reason, 'token expired');
    });

    test('parses a domain event envelope', () {
      final message = parseIncomingMessage('''
        {
          "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
          "type": "grades.published",
          "room": "school:123:role:STUDENT",
          "payload": {"a": 1},
          "publishedAt": "2026-07-09T12:00:00.000Z"
        }
      ''');
      expect(message, isA<EventEnvelope>());
    });

    test('drops malformed JSON', () {
      expect(parseIncomingMessage('{not json'), isNull);
    });

    test('drops JSON that is not an object', () {
      expect(parseIncomingMessage('[1,2,3]'), isNull);
    });

    test('drops an unknown message shape', () {
      expect(parseIncomingMessage('{"type":"system.unknown"}'), isNull);
    });
  });

  group('ClientMessage.toJson', () {
    test('join', () {
      expect(const JoinRoomMessage('school:123').toJson(), {
        'type': 'join',
        'room': 'school:123',
      });
    });

    test('leave', () {
      expect(const LeaveRoomMessage('school:123').toJson(), {
        'type': 'leave',
        'room': 'school:123',
      });
    });
  });
}
